R/container.R

```diff
@@ -0,0 +1,174 @@
+#' container()
+#' @export
+container <- function(mode = "unknown", type = "unknown", outcome = character(0),
+                      estimate = character(0), probabilities = character(0),
+                      time = character(0), call = rlang::current_env()) {
+  dat <-
+    list(
+      outcome = outcome,
+      type = type,
+      estimate = estimate,
+      probabilities = probabilities,
+      time = time
+    )
+  new_container(
+    mode,
+    type,
+    operations = list(),
+    columns = dat,
+    ptype = tibble::tibble(),
+    call = call
+  )
+}
+
+new_container <- function(mode, type, operations, columns, ptype, call) {
+  mode <- rlang::arg_match0(mode, c("unknown", "regression", "classification", "censored regression"))
+
+  if ( mode == "regression" ) {
+    type <- "regression"
+  }
+
+  type <- rlang::arg_match0(type, c("unknown", "regression", "binary", "multiclass"))
+
+  if ( !is.list(operations) ) {
+    cli::cli_abort("The {.arg operations} argument should be a list.", call = call)
+  }
+
+  is_oper <- purrr::map_lgl(operations, ~ inherits(.x, "operation"))
+  if ( length(is_oper) > 0 & !any(is_oper) ) {
+    bad_oper <- names(is_oper)[!is_oper]
+    cli::cli_abort("The following {.arg operations} do not have the class \\
+                   {.val operation}: {bad_oper}.", call = call)
+  }
+
+  # validate operation order and check duplicates
+  validate_oper_order(operations, mode, call)
+
+
+  # check columns
+
+  res <- list(mode = mode, type = type, operations = operations,
+              columns = columns, ptype = ptype)
+  class(res) <- "container"
+  res
+}
+
+#' @export
+print.container <- function(x, ...) {
+  # todo emulate Emil's recipe printing
+
+  num_op <- length(x$operations)
+  cli::cli_inform("{x$type} post-processing object with {num_op} operation{?s}")
+
+  if (num_op > 0) {
+    cat("\n")
+    res <- purrr::map(x$operations, ~ print(.x))
+  }
+
+  invisible(x)
+}
+
+
+# ------------------------------------------------------------------------------
+
+#' @export
+fit.container <- function(object, .data, outcome, estimate, probabilities = c(),
+                          time = c(), call = rlang::current_env(), ...) {
+
+  # ------------------------------------------------------------------------------
+  # set columns via tidyselect
+
+  dat <- list()
+  dat$outcome <- names(tidyselect::eval_select(rlang::enquo(outcome), .data))
+  dat$estimate <- names(tidyselect::eval_select(rlang::enquo(estimate), .data))
+
+  probabilities <- tidyselect::eval_select(rlang::enquo(probabilities), .data)
+  if (length(probabilities) > 0) {
+    dat$probabilities <- names(probabilities)
+  } else {
+    dat$probabilities <- character(0)
+  }
+
+  time <- tidyselect::eval_select(rlang::enquo(time), .data)
+  if (length(time) > 0) {
+    dat$time <- names(time)
+  }    else {
+    dat$time <- character(0)
+  }
+
+  .data <- .data[, names(.data) %in% unlist(dat)]
+  .data <- tibble::as_tibble(.data)
+  ptype <- .data[0,]
```

Comment:
Just a note for future us: I considered adding methods for applicability scores. We'd need to update the `ptype` to include the columns of the training set predictors (the data would be an argument to such functions).

Reply:
#4